### PR TITLE
Replace build layouts with Launcher_PORTABLE cmake flag

### PR DESCRIPTION
**Pull request by txtsd**
~~Allows Windows to start using %APPDATA% to store PolyMC data instead of the `cwd`.~~

~~One step closer to an installer.~~

Okay `Launcher_PORTABLE` is now a build flag. It is `ON` by default.

Data path defaults to bin dir by default, except on macOS.
On Windows and Linux, if `LAUNCHER_PORTABLE=OFF`, then `QStandardPaths::AppDataLocation` is used instead.
EDIT: On any other OS, not just Windows and Linux.

This replaces the build layouts that were in place earlier, so external build scripts will have to be updated. I've taken care of BUILD.md and the nix thing.

EDIT: Also, the linux system directory was previously hardcoded to use `polymc` in `XDG_DATA_HOME`. However, `QStandardPaths::AppDataLocation` outputs `PolyMC`, which is how it is —how it should be— on other OSes. Looking for inputs on how to handle this.

EDIT: Also reverts the `m_rootPath` from #146

### CODE REVIEW

**Comment by DioEgizio** [comment]
cool

**Comment by kloenk** [comment]
Isn't this now the same as on macOS? Can we simplify that, and put it into a common codepath?

**Comment by LennyMcLennington** [comment]
This should be an option at build time instead of forcing it on everyone, it doesn't make sense if it's released as a portable app but you're changing stuff with the assumption that it's installed to the system.

File: launcher/Application.cpp [hunk 1]
```diff
@@ -529,7 +533,8 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
         QDir foo(FS::PathCombine(binPath, ".."));
         m_rootPath = foo.absolutePath();
 #elif defined(Q_OS_WIN32)
-        m_rootPath = binPath;
+        QDir foo(FS::PathCombine(QStandardPaths::writableLocation(QStandardPaths::AppDataLocation), ".."));
```
Comment (Scrumplex) [inline launcher/Application.cpp]: Something is not quite right here, if you look a few lines above it says "Root path is used for updates."

**Review by d-513 — CHANGES_REQUESTED** [review]
Make a build flag, so that it can still be used portable

**Comment by txtsd** [comment]
> Make a build flag, so that it can still be used portable

Sounds fair. Should I do the same for Linux?

For macOS, I assume writing within the .app not useful?

**Comment by DioEgizio** [comment]
> > Make a build flag, so that it can still be used portable
> 
> Sounds fair. Should I do the same for Linux?
> 
> For macOS, I assume writing within the .app not useful?

You should never write things directly in the .app. No idea on why MMC did

**Comment by txtsd** [comment]
> This should be an option at build time instead of forcing it on everyone, it doesn't make sense if it's released as a portable app but you're changing stuff with the assumption that it's installed to the system.

It's not released as a portable app though, that's just what I named those builds in the beginning because none of them used data dirs by default. peterix did not want to use standard data locations that's how he designed it.

Shouldn't we use standard data paths by default?

File: launcher/Application.cpp (hunk 1 above)
Comment (txtsd) [reply launcher/Application.cpp]: Does MultiMC auto update?

**Comment by d-513** [comment]
when using installer use correct paths
portable should write in its own dir 

having portable builds is useful still, we shouldn’t remove it as an option

**Comment by d-513** [comment]
And we should use portable by default, makes it easier to test, installer with correct dirs on releases

File: launcher/Application.cpp (hunk 1 above)
Comment (LennyMcLennington) [reply launcher/Application.cpp]: No but it has an updater

Comment (txtsd) [reply launcher/Application.cpp]: That's MultiMC's updater which it uses from the launch script which we don't use, so it should be fine.

Comment (txtsd) [reply launcher/Application.cpp]: I'll revert these `m_rootPath` changes. #146 from earlier changed it too.

**Comment by LennyMcLennington** [comment]
> > Make a build flag, so that it can still be used portable
> 
> Sounds fair. Should I do the same for Linux?

Linux already writes to ~/.local/share/polymc on lin-system layout or to the directory that the executable is in on lin-nodeps layout?

**Comment by txtsd** [comment]
> > > Make a build flag, so that it can still be used portable
> > 
> > Sounds fair. Should I do the same for Linux?
> 
> Linux already writes to ~/.local/share/polymc on lin-system layout or to the directory that the executable is in on lin-nodeps layout?

I'm gonna get rid of the layouts in favor of a boolean build flag. It will switch Windows and Linux between portable and data dir.

Default will be portable.

**Comment by ghost** [comment]
This will require changes to AUR and MPR, please let me know when this is merged so I can make the necessary changes

File: BUILD.md [hunk 2]
```diff
@@ -250,10 +250,11 @@ zlib1.dll
...
 3. Make a build directory, and change directory to the directory and do `cmake -G "MinGW Makefiles" -DCMAKE_INSTALL_PREFIX=C:\Path\that\makes\sense\for\you`. By default, it will install to C:\Program Files (x86), which you might not want, if you want a local installation. If you want to install it to that directory, make sure to run the command window as administrator.
-4. Do `mingw32-make -jX`, where X is the number of cores your CPU has plus one.
-5. Now to wait for it to compile. This could take some time. Hopefully it compiles properly.
-6. Run the command `mingw32-make install`, and it should install PolyMC, to whatever the `-DCMAKE_INSTALL_PREFIX` was.
-7. In most cases, whenever compiling, the OpenSSL dll's aren't put into the directory to where PolyMC installs, meaning you cannot log in. The best way to fix this is just to do `copy C:\OpenSSL-Win32\*.dll C:\Where\you\installed\PolyMC\to`. This should copy the required OpenSSL dll's to log in.
+4. If you want PolyMC to store its data in `%APPDATA%`, append `-DLAUNCHER_PORTABLE=OFF` to the previous command.
+5. Do `mingw32-make -jX`, where X is the number of cores your CPU has plus one.
```
Comment (ghost) [inline BUILD.md]: just do `-j$(nproc)`

File: launcher/Application.cpp [hunk 3]
```diff
@@ -306,16 +306,11 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
...
     else
     {
-#ifdef LAUNCHER_LINUX_DATADIR
-        QString xdgDataHome = QFile::decodeName(qgetenv("XDG_DATA_HOME"));
-        if (xdgDataHome.isEmpty())
-            xdgDataHome = QDir::homePath() + QLatin1String("/.local/share");
-        dataPath = xdgDataHome + "/polymc";
-        adjustedBy += "XDG standard " + dataPath;
-#elif defined(Q_OS_MAC)
+        // qDebug() << LAUNCHER_PORTABLE;
```
Comment (ghost) [inline launcher/Application.cpp]: z

**Comment by txtsd** [comment]
Okay `Launcher_PORTABLE` is now a build flag. It is `ON` by default.

Data path defaults to bin dir by default, except on macOS.
On Windows and Linux, if `LAUNCHER_PORTABLE=OFF`, then `QStandardPaths::AppDataLocation` is used instead.
EDIT: On any other OS, not just Windows and Linux.

This replaces the build layouts that were in place earlier, so external build scripts will have to be updated. I've taken care of BUILD.md and the nix thing.

EDIT: Also, the linux system directory was previously hardcoded to use `polymc` in `XDG_DATA_HOME`. However, `QStandardPaths::AppDataLocation` outputs `PolyMC`, which is how it is —how it should be— on other OSes. Looking for inputs on how to handle this.

EDIT: Also reverts the `m_rootPath` from #146

**Review by ghost — DISMISSED** [review]

File: launcher/Application.cpp (hunk 3 above)
Comment (txtsd) [reply launcher/Application.cpp]: Derp. Good catch.

File: BUILD.md (hunk 2 above)
Comment (txtsd) [reply BUILD.md]: I didn't change it. Just the line number. Should I still make that change in this PR?

Comment (ghost) [reply BUILD.md]: yes

File: CMakeLists.txt [hunk 4]
```diff
@@ -146,25 +146,11 @@ add_subdirectory(program_info)
...
-    elseif(WIN32)
-        set(Launcher_LAYOUT_REAL "win-bundle")
-    else()
-        message(FATAL_ERROR "Cannot choose a sensible install layout for your platform.")
-    endif()
-else()
-    set(Launcher_LAYOUT_REAL ${Launcher_LAYOUT})
-endif()
+# Install the build results according to platform
+set(LAUNCHER_PORTABLE 1 CACHE BOOL "The type of installation (Portable or System)")
```
Comment (Scrumplex) [inline CMakeLists.txt]: I would prefer if this was spelled `Launcher_PORTABLE`

Comment (ghost) [reply CMakeLists.txt]: thats standard for other flags, please do so

Comment (txtsd) [reply CMakeLists.txt]: I'd rather `s/Launcher_/LAUNCHER_/g` to keep all variables consistently named. What do you think?

File: launcher/Application.cpp [hunk 5]
```diff
@@ -306,16 +306,11 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
...
+#if !LAUNCHER_PORTABLE || defined(Q_OS_MAC)
         QDir foo(FS::PathCombine(QStandardPaths::writableLocation(QStandardPaths::AppDataLocation), ".."));
```
Comment (Scrumplex) [inline launcher/Application.cpp:298]: This changes the data path on Linux from `.local/share/polymc` to `.local/share/PolyMC`.

**Review by Scrumplex — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 4 above)
Comment (Scrumplex) [reply CMakeLists.txt]: I would prefer to do `s/Launcher_/POLYMC_/g`, but not in this PR. We can do it in a different PR.

File: launcher/Application.cpp (hunk 5 above)
Comment (txtsd) [reply launcher/Application.cpp:298]: Oh yes, completely forgot to mention that in my changes comment.

This is an unfortunate change, however the folder should've been named correctly in the first place.

How do we handle this?

Comment (ghost) [reply launcher/Application.cpp:298]: keep it as polymc to stay consistent and avoid breaking stuff

Comment (Scrumplex) [reply launcher/Application.cpp:298]: Maybe we can use both paths, but prefer PolyMC. Perhaps tell the user that they should rename the directory when we detect the old name. (I don't like applications moving their data directories without me knowing)

Comment (Scrumplex) [reply launcher/Application.cpp:298]: Let me prepare a patch

Comment (txtsd) [reply launcher/Application.cpp:298]: That's unnecessary cruft code to maintain backward compatibility with a previously incorrectly set value, imo.

File: CMakeLists.txt (hunk 4 above)
Comment (txtsd) [reply CMakeLists.txt]: Okay then I'll change it to `Launcher_` for now.

File: launcher/Application.cpp (hunk 5 above)
Comment (Scrumplex) [reply launcher/Application.cpp:298]: But we want a smooth upgrade path for users. Why annoy them with seemingly unnecessary changes?

Comment (txtsd) [reply launcher/Application.cpp:298]: > Let me prepare a patch

You can push right into this PR.

Comment (txtsd) [reply launcher/Application.cpp:298]: > But we want a smooth upgrade path for users. Why annoy them with seemingly unnecessary changes?

Yea that's fair.

File: launcher/Application.cpp [hunk 6]
```diff
@@ -306,16 +306,20 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
...
-#elif defined(Q_OS_MAC)
+#if !Launcher_PORTABLE || defined(Q_OS_MAC)
         QDir foo(FS::PathCombine(QStandardPaths::writableLocation(QStandardPaths::AppDataLocation), ".."));
         dataPath = foo.absolutePath();
-        adjustedBy += "Fallback to special Mac location " + dataPath;
+        adjustedBy += dataPath;
+
+#ifdef Q_OS_LINUX
+        // TODO: this should be removed in a future version
+        // TODO: provide a migration path similar to macOS migration
```
Comment (Scrumplex) [inline launcher/Application.cpp:304]: The migration path should be added in a separate PR I think. This should be fine for now.

**Review by Scrumplex — APPROVED** [review]

File: launcher/Application.cpp (hunk 6 above)
Comment (txtsd) [reply launcher/Application.cpp:304]: Looks good!

**Comment by Scrumplex** [comment]
IMO this works as expected on Linux. Need confirmation from Windows and macOS ppl.

**Comment by ghost** [comment]
I can test this on both sometime later. Maybe today, but no promises.

**Comment by kloenk** [comment]
Likely unrelated, but does not build on x86_64-darwin with nix.

<details>
<summary>Last build log lines</summary>
```
CMake Error at launcher/cmake_install.cmake:55 (file):
  file INSTALL cannot find
  "/private/tmp/nix-build-qtbase-5.12.10.drv-0/qtbase-everywhere-src-5.12.10/$(out)/$(qtPluginPrefix)/imageformats":
  No such file or directory.
Call Stack (most recent call first):
  cmake_install.cmake:62 (include)


FAILED: CMakeFiles/install.util
cd /tmp/nix-build-polymc-nightly.drv-0/source/build && /nix/store/dvbbh8zf4bmw106rfvmnhfwqbihyyz6j-cmake-3.22.1/bin/cmake -P cmake_install.cmake
```
</details>

**Comment by kloenk** [comment]
nix build on nixos `x86_64-linux` works, but can't test the resulting binary, as I'm missing a window server.

**Comment by txtsd** [comment]
> Likely unrelated, but does not build on x86_64-darwin with nix.
> Last build log lines
> CMake Error at launcher/cmake_install.cmake:55 (file): file INSTALL cannot find "/private/tmp/nix-build-qtbase-5.12.10.drv-0/qtbase-everywhere-src-5.12.10/$(out)/$(qtPluginPrefix)/imageformats": No such file or directory. Call Stack (most recent call first): cmake_install.cmake:62 (include)
> 
> FAILED: CMakeFiles/install.util cd /tmp/nix-build-polymc-nightly.drv-0/source/build && /nix/store/dvbbh8zf4bmw106rfvmnhfwqbihyyz6j-cmake-3.22.1/bin/cmake -P cmake_install.cmake

How did you call cmake?

**Comment by kloenk** [comment]
> How did you call cmake?

``` 
nix build <flakeref>
```

**Comment by txtsd** [comment]
Sorry, not familiar with nix stuff.

Nothing regarding the Qt plugins changed, so you shouldn't be getting that error. Is it failing only on this PR? Is it working on develop?

EDIT: If you leave an empty line after the `<details>`, you can wrap the code in ```

**Review by timoreo22 — APPROVED** [review]
Looks good !

**Review by d-513 — APPROVED** [review]
All good/already mentioned by others

**Comment by txtsd** [comment]
Why is this still blocked when the changes requested were resolved?

**Comment by Scrumplex** [comment]
There you go

**Comment by Scrumplex** [comment]
@LennyMcLennington please review this. It's a larger change and I want you to look at this as well.

File: CMakeLists.txt [hunk 7]
```diff
@@ -195,47 +181,40 @@ if(Launcher_LAYOUT_REAL STREQUAL "mac-bundle")
...
-elseif(Launcher_LAYOUT_REAL STREQUAL "lin-system")
-    set(Launcher_BINARY_DEST_DIR "bin" CACHE STRING "Path to the binary directory")
-    set(Launcher_LIBRARY_DEST_DIR "lib${LIB_SUFFIX}" CACHE STRING "Path to the library directory")
-    set(Launcher_SHARE_DEST_DIR "share/polymc" CACHE STRING "Path to the shared data directory")
-    set(JARS_DEST_DIR "${Launcher_SHARE_DEST_DIR}/jars")
-    set(Launcher_DESKTOP_DEST_DIR "share/applications" CACHE STRING "Path to the desktop file directory")
-    set(Launcher_METAINFO_DEST_DIR "share/metainfo" CACHE STRING "Path to the metainfo directory")
-    set(Launcher_ICON_DEST_DIR "share/icons/hicolor/scalable/apps" CACHE STRING "Path to the scalable icon directory")
+        # launcher/Application.cpp will use this value
+        set(Launcher_APP_BINARY_DEFS "-DLauncher_PORTABLE=1")
```
Comment (LennyMcLennington) [inline CMakeLists.txt]:
```suggestion
        set(Launcher_APP_BINARY_DEFS "-DLAUNCHER_PORTABLE")
```

File: CMakeLists.txt [hunk 8]
```diff
@@ -195,47 +181,40 @@ if(Launcher_LAYOUT_REAL STREQUAL "mac-bundle")
...
+    else()
+        set(LIBRARY_DEST_DIR "lib${LIB_SUFFIX}")
+        set(JARS_DEST_DIR "share/jars")
+        set(LAUNCHER_DESKTOP_DEST_DIR "share/applications" CACHE STRING "Path to the desktop file directory")
+        set(LAUNCHER_METAINFO_DEST_DIR "share/metainfo" CACHE STRING "Path to the metainfo directory")
+        set(LAUNCHER_ICON_DEST_DIR "share/icons/hicolor/scalable/apps" CACHE STRING "Path to the scalable icon directory")
 
-    MESSAGE(STATUS "Compiling for linux system with ${Launcher_SHARE_DEST_DIR} and LAUNCHER_LINUX_DATADIR")
-    SET(Launcher_APP_BINARY_DEFS "-DMULTIMC_JARS_LOCATION=${CMAKE_INSTALL_PREFIX}/${JARS_DEST_DIR}" "-DLAUNCHER_LINUX_DATADIR")
+        set(Launcher_APP_BINARY_DEFS "-DMULTIMC_JARS_LOCATION=${CMAKE_INSTALL_PREFIX}/${JARS_DEST_DIR}" "-DLauncher_PORTABLE=0")
```
Comment (LennyMcLennington) [inline CMakeLists.txt]:
```suggestion
        set(Launcher_APP_BINARY_DEFS "-DMULTIMC_JARS_LOCATION=${CMAKE_INSTALL_PREFIX}/${JARS_DEST_DIR}")
```

File: launcher/Application.cpp [hunk 9]
```diff
@@ -294,16 +294,20 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
...
     else
     {
-#ifdef LAUNCHER_LINUX_DATADIR
-        QString xdgDataHome = QFile::decodeName(qgetenv("XDG_DATA_HOME"));
-        if (xdgDataHome.isEmpty())
-            xdgDataHome = QDir::homePath() + QLatin1String("/.local/share");
-        dataPath = xdgDataHome + "/polymc";
-        adjustedBy += "XDG standard " + dataPath;
-#elif defined(Q_OS_MAC)
+#if !Launcher_PORTABLE || defined(Q_OS_MAC)
```
Comment (LennyMcLennington) [inline launcher/Application.cpp]:
```suggestion
#if !defined(LAUNCHER_PORTABLE) || defined(Q_OS_MAC)
```

**Review by LennyMcLennington — DISMISSED** [review]
Using the definition itself as a flag makes more sense to me, what does everyone else think?

File: CMakeLists.txt (hunk 7 above)
Comment (d-513) [reply CMakeLists.txt]: other settings use DLauncher

Comment (LennyMcLennington) [reply CMakeLists.txt]: @dada513 You are mistaking macro definitions with CMake option definitions. The macros use `LAUNCHER_` and the CMake options use `Launcher_`. It is a very common naming convention to name preprocessor macros with all upper-case letters, you can see that with pretty much every other macro that is used. It's also conventional to use the fact that a macro is defined to test for a flag, rather than the value of the macro, which is why I removed the `=1` part and changed the test to test for `!defined(LAUNCHER_PORTABLE)`.

Comment (txtsd) [reply CMakeLists.txt]: @LennyMcLennington `Application.cpp` needs `Launcher_PORTABLE=1` or it doesn't work. I couldn't find a better way for it to check the boolean. Did you see if the logic works well after changing it?

Also, we decided to change the flags to `POLYMC_*` in a future PR, which is why I left it lowercase to fit in with the other flags.

Comment (LennyMcLennington) [reply CMakeLists.txt]: > @LennyMcLennington `Application.cpp` needs `Launcher_PORTABLE=1` or it doesn't work. I couldn't find a better way for it to check the boolean. Did you see if the logic works well after changing it?

I know C and C++ so I am pretty confident that it will work. Read the change I suggested in Application.cpp.

> Also, we decided to change the flags to `POLYMC_*` in a future PR, which is why I left it lowercase to fit in with the other flags.

Changing the CMake flags prefixes to `PolyMC_` and changing the preprocessor macros to `POLYMC_` has literally nothing to do with the case of naming conventions for C Preprocessor macros.

I am assuming that the change you said you've decided on still keeps the exact same naming conventions and just changes the word from 'Launcher' to 'PolyMC', If it doesn't then I have a problem with it.
